chain/checkpointing/util.go

```diff
@@ -0,0 +1,288 @@
+	tweaked_value := HashTweakedValue(internal_pubkey, merkle_root)
+
+	tweaked_pubkey := ApplyTweakToPublicKeyTaproot(internal_pubkey, tweaked_value)
+	return tweaked_pubkey
+}
+
+func AddTaprootScriptToWallet(taprootScript string) bool {
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"importaddress\", \"params\": [\"" + taprootScript + "\", \"\", true]}"
+	result := jsonRPC(payload)
+
+	if result["error"] == nil {
+		return true
+	}
+
+	err := result["error"].(map[string]interface{})
+	if err["code"].(float64) == -4 {
+		// Particular case where we are already in the process of adding the key
+		// because we are using 1 bitcoin node for all
+		return true
+	}
+
+	return false
+}
+
+func GetTaprootScript(pubkey []byte) string {
+	return "5120" + hex.EncodeToString(pubkey)
+}
+
+// Temporary
+func BitcoindGetWalletAddress() string {
+	// Create wallet
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"createwallet\", \"params\": [\"wow\"]}"
+	_ = jsonRPC(payload)
+	// We don't cehck error here
+
+	//Get new address
+	payload = "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"getnewaddress\", \"params\": []}"
+
+	result := jsonRPC(payload)
+	address := fmt.Sprintf("%v", result["result"])
+	return address
+}
+
+func WalletGetTxidFromAddress(taprootAddress string) (string, error) {
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"listtransactions\", \"params\": [\"*\", 500000000, 0, true]}"
+	result := jsonRPC(payload)
+	list := result["result"].([]interface{})
+	for _, item := range list {
+		item_map := item.(map[string]interface{})
+		if item_map["address"] == taprootAddress {
+			txid := item_map["txid"].(string)
+			return txid, nil
+		}
+	}
+	return "", errors.New("did not find checkpoint")
+}
+
+func BitcoindPing() bool {
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"ping\", \"params\": []}"
+	result := jsonRPC(payload)
+	return result != nil
+}
+
+func PrepareWitnessRawTransaction(rawtx string, sig []byte) string {
+	wtx := rawtx[:4*2] + "00" + "01" + rawtx[4*2:len(rawtx)-4*2] + "01" + "40" + hex.EncodeToString(sig) + rawtx[len(rawtx)-4*2:]
+	fmt.Println("Raw transaction signed :", wtx)
+
+	return wtx
+}
+
+func ParseUnspentTxOut(utxo []byte) (amount, script []byte) {
+	return utxo[0:8], utxo[9:]
+}
+
+func GetTxOut(txid string, index int) (float64, []byte) {
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"gettxout\", \"params\": [\"" + txid + "\", " + strconv.Itoa(index) + "]}"
+	result := jsonRPC(payload)
+	fmt.Println(result)
+	if result == nil {
+		panic("cant retrieve previous transaction")
+	}
+	taprootTxOut := result["result"].(map[string]interface{})
+	scriptPubkey := taprootTxOut["scriptPubKey"].(map[string]interface{})
+	scriptPubkeyBytes, _ := hex.DecodeString(scriptPubkey["hex"].(string))
+
+	return taprootTxOut["value"].(float64), scriptPubkeyBytes
+}
+
+func jsonRPC(payload string) map[string]interface{} {
```

Comment:
Why not returning an error in this function?

Reply:
Yes, it should return an error. Fixing this.